[JENKINS-27058] Clarify that remoteFS can be relative if you know what y...

stephenc opened the pull request:
...ou are doing
- Get the absolute path on connect and use that while the connection is on-line (which is ok as when the connection is not online a null was returned from getRootPath() anyway)

stephenc committed:
[JENKINS-27058] Clarify that remoteFS can be relative if you know what you are doing

- Get the absolute path on connect and use that while the connection is on-line (which is ok as when the connection is not online a null was returned from getRootPath() anyway)

stephenc commented:
To replace https://github.com/jenkinsci/jenkins/pull/1578 @reviewbybees

jtnord commented on core/src/main/java/hudson/model/Slave.java:
```diff
@@ -121,14 +126,14 @@
     private String label="";
-    
+
     private /*almost final*/ DescribableList<NodeProperty<?>,NodePropertyDescriptor> nodeProperties = new DescribableList<NodeProperty<?>,NodePropertyDescriptor>(Jenkins.getInstance());
```
people in greenhouses shouldn't throw stones :-)

stephenc replied:
I comment on other's whitespace changes iff they have been especially nit-picky in their own review comments

jtnord commented:
:+1: ignoring whitespace unrelated changes.

jglick commented on core/src/main/java/hudson/model/Slave.java:
```diff
@@ -425,7 +437,7 @@ public FormValidation doCheckRemoteFS(@QueryParameter String value) throws IOExc
                 // Unix-looking path that doesn't start with '/'
                 // TODO: detect Windows-looking relative path
-                return FormValidation.error(Messages.Slave_the_remote_root_must_be_an_absolute_path());
+                return FormValidation.error(Messages.Slave_Remote_Relative_Path_Warning());
```
Not `.warning`?

jglick commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -410,6 +412,11 @@ public long getResourceLoadingTime() throws IOException, InterruptedException {
     }
 
+    @CheckForNull
+    public String getAbsoluteRemoteFs() {
```
Either `@Restricted` or `@since` or not `public`.

jglick commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -481,11 +488,16 @@ public void onClosed(Channel c, IOException cause) {
-        FilePath root = new FilePath(channel,remoteFs);
+
+        String absoluteRemoteFs = node.getRemoteFS();
+        if (!absoluteRemoteFs.startsWith("\\") && !absoluteRemoteFs.startsWith("/")) {
```
What about `C:\absolute`?

jglick commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -634,9 +647,13 @@ public RetentionStrategy getRetentionStrategy() {
-        channel = null;
-        isUnix = null;
+        Channel c;
+        synchronized (channelLock) {
```
How is this related to the rest of the PR?

reviewbybees replied:
We need to clear the absolute path atomically with closing the channel, this just adds isUnix into that atomic operation

Sent from my iPhone

> On 16 Mar 2015, at 19:13, Jesse Glick <notifications@github.com> wrote:
> 
> In core/src/main/java/hudson/slaves/SlaveComputer.java:
> 
> > @@ -634,9 +647,13 @@ public RetentionStrategy getRetentionStrategy() {
> >       */
> >      private void closeChannel() {
> >          // TODO: race condition between this and the setChannel method.
> > -        Channel c = channel;
> > -        channel = null;
> > -        isUnix = null;
> > +        Channel c;
> > +        synchronized (channelLock) {
> How is this related to the rest of the PR?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/jenkins/pull/1606/r26517858%40github.com.

jglick commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -708,6 +725,18 @@ public Boolean call() throws IOException {
         }
     }
 
+    private static final class AbsolutePath extends MasterToSlaveCallable<String,IOException> {
```
`serialVersionUID`

jglick commented on core/src/main/resources/hudson/model/Messages.properties:
```diff
@@ -244,7 +244,9 @@ Slave.Launching={0} Launching slave agent
-Slave.the_remote_root_must_be_an_absolute_path=The remote root must be an absolute path.
+Slave.Remote.Relative.Path.Warning=Are you sure you want to use a relative path for the FS root? Note that relative \
+  paths require that you can assure that the selected launcher provides a consistent current working directory. Using \
+  an absolute path is strongly encouraged and highly recommended.
```
`strongly encouraged and highly recommended` is redundant.

jglick commented:
Hard to review given the excessive whitespace changes, but :+1: overall insofar as I understand the implications.

daniel-beck commented on core/src/main/resources/hudson/model/Messages.properties:
```diff
@@ -244,7 +244,9 @@ Slave.Launching={0} Launching slave agent
 Slave.Terminated={0} slave agent was terminated
-Slave.the_remote_root_must_be_an_absolute_path=The remote root must be an absolute path.
+Slave.Remote.Relative.Path.Warning=Are you sure you want to use a relative path for the FS root? Note that relative \
+  paths require that you can assure that the selected launcher provides a consistent current working directory. Using \
```
What does this mean?

jglick replied:
I think Stephen is trying to say something about cloudy slaves that share a common filesystem despite apparently different launch details, but I am not really sure either. Deserves a clearer explanation.

reviewbybees replied:
If you use JNLP to launch, the CWD depends on where you run the command, so (apart from our jnlp cloudy slaves) you probably should not use a relative path with jnlp (unless using jnlp cloudy slaves where it is very very useful)

With ssh the CWD will typically be the user's home directory and this is consistent so relative paths are very safe with ssh

Sent from my iPhone

> On 17 Mar 2015, at 13:56, Jesse Glick <notifications@github.com> wrote:
> 
> In core/src/main/resources/hudson/model/Messages.properties:
> 
> > @@ -244,7 +244,9 @@ Slave.Launching={0} Launching slave agent
> >  Slave.Network.Mounted.File.System.Warning=Are you sure you want to use network mounted file system for FS root? Note that this directory does not need to be visible to the master.
> >  Slave.Remote.Director.Mandatory=Remote directory is mandatory
> >  Slave.Terminated={0} slave agent was terminated
> > -Slave.the_remote_root_must_be_an_absolute_path=The remote root must be an absolute path.
> > +Slave.Remote.Relative.Path.Warning=Are you sure you want to use a relative path for the FS root? Note that relative \
> > +  paths require that you can assure that the selected launcher provides a consistent current working directory. Using \
> I think Stephen is trying to say something about cloudy slaves that share a common filesystem despite apparently different launch details, but I am not really sure either. Deserves a clearer explanation.
> 
> —
> Reply to this email directly or view it on GitHub.
> 
> -- 
> You received this message because you are subscribed to the Google Groups "engineering-code-reviews" group.
> To unsubscribe from this group and stop receiving emails from it, send an email to engineering-code-reviews+unsubscribe@cloudbees.com.
> To post to this group, send email to engineering-code-reviews@cloudbees.com.
> To view this discussion on the web visit https://groups.google.com/a/cloudbees.com/d/msgid/engineering-code-reviews/jenkinsci/jenkins/pull/1606/r26573715%40github.com.

daniel-beck replied:
This should be part of the documentation help text.

stephenc committed:
Doeth! I meant to have this as warning

stephenc committed:
Handle mostly-absolute paths on windows that use a drive letter

- of course drive mappings are user scoped and can change from user session to user session, so calling them absolute is a bit of a joke... none the less, for our purposes they are absolute enough

stephenc committed:
Define a consistent serialVersionUID

- Ignoring all the other callables in this class that are missing a serialVersionUID

stephenc committed:
Simplify english

stephenc committed:
Augment help text to explain when relative paths are OK

stephenc commented:
@daniel-beck @jglick I think I have addressed your issues

stephenc committed:
Address @daniel-beck's comments

jglick commented on core/src/main/java/hudson/slaves/SlaveComputer.java:
```diff
@@ -537,6 +559,22 @@ public void onClosed(Channel c, IOException cause) {
 
+    /*package*/ static boolean isRelativePath(String path) {
+        if (path.startsWith("/"))
+            return false;
```
Debatable whether you want to consider `\Windows\Temp\slave` absolute or relative. Technically it is drive-relative.

stephenc replied:
`\windows\temp\slave` is relative, `/windows/temp/slave` is fine to consider as absolute as you are indicating that you think it is unix and the `isUnix` test will warn you about the path being for a different OS

jglick commented on core/src/test/java/hudson/slaves/SlaveComputerTest.java:
```diff
@@ -0,0 +1,65 @@
+        assertThat(".\\foo\\bar", aRelativePath());
+        assertThat(".\\foo\\bar\\", aRelativePath());
+        assertThat("\\\\foo", aRelativePath());
+        assertThat("\\\\foo\\", not(aRelativePath()));
```
FWIW the share name is a mandatory part of a UNC path, so `\\server\share\localpath`. I doubt the distinction matters here.

stephenc replied:
I know but I wanted a faster test and this is better than anything else we had... it can be improved if somebody feels the need ;-)

jglick commented on core/src/test/java/hudson/slaves/SlaveComputerTest.java:
```diff
@@ -0,0 +1,65 @@
+        return new RelativePathMatcher();
+    }
+
+    private static class RelativePathMatcher extends BaseMatcher<String> {
```
Does Hamcrest actually make your life easier?

stephenc replied:
I prefer reading it this way, so yes

jglick commented:
:+1:

jglick commented:
Recommend review from @recampbell since he encountered the issue to begin with.

stephenc commented:
I'll wait for @recampbell so...

recampbell commented:
:+1:

stephenc committed:
Move isRelativePath to Util and use from hudson.model.Slave's descriptor

stephenc commented:
Jenkins found a failing test with the changes for @jglick, so some refactoring to reduce code duplication

stephenc committed:
don't call this local variable absolute as it maybe relative... it will be absolute when we are finished mind!

stephenc commented:
Test failures in most recent build are irrelevant

KostyaSha commented:
Why all whitespace changes were done in one commit with real changes?

stephenc commented:
IntelliJ screwed me over and refused to unscrew

KostyaSha commented:
On left side of the code (near line numbers) there is a grey colour that shows that something was changed. If you press on it it allows to undo it. Also there was somewhere diff button that allows compare current vs previous changes and merge them.

jglick commented:
At worst,

1. `git checkout mybranch`
1. `git merge master`
1. `git diff master > /tmp/x.diff`
1. edit `/tmp/x.diff` and delete all useful changes
1. `git apply -R /tmp/x.diff`
1. `git commit -A -m 'Writing ten times on the blackboard, I will not commit gratuitous WS changes again.'`

:-)

KostyaSha commented:
@jglick just `git add -i ` then `patch` and skip WS hanks.

stephenc commented:
@KostyaSha so the issue I had is that I set intellij to leave whitespace alone on unmodified lines, but then every time I did the grey trick it would revert it on save... Normally I don't have this issue but I think it was seeing me "undoing" the whitespace changes as changing those lines and thus they were up for whitespace fixing... lather rinse repeat

KostyaSha commented:
I think we should define settings and document on wiki, i also spending much time doing PRs without whitespaces :(

KostyaSha commented:
![screenshot 2015-03-23 21 46 12](https://cloud.githubusercontent.com/assets/231611/6787898/dbc40a2e-d1a6-11e4-8df9-0c6ad0e9909e.png)
Seems this IDEA 14 feature doesn't work right. It acts like "All"

KostyaSha commented:
Reproduces, open SafeTimerTask.class, add new line, alt+tab window, back... May this be caused by some git configuration?

stephenc commented:
That's the exact issue I had. It used to work in IDEA 13... ffs

KostyaSha commented:
Filled bug request to their support because i have no other ideas.

KostyaSha commented:
@stephenc support replied, IDEA supports editor config https://youtrack.jetbrains.com/issue/IDEA-130042#comment=27-804824 that was put by this https://github.com/jenkinsci/jenkins/pull/1310  https://github.com/jenkinsci/jenkins/blob/master/.editorconfig#L16 
As i see this format sucks on options and my proposal to remove it.

jtnord commented:
it's also has no support for eclipse :-)

Note: even eclipse for the same eclipse format config will generate
different formats between versions as bugs get fixed/introduced and new
features are added.

On 24 March 2015 at 08:22, Kanstantsin Shautsou <notifications@github.com>
wrote:

> @stephenc <https://github.com/stephenc> support replied, IDEA supports
> editor config
> https://youtrack.jetbrains.com/issue/IDEA-130042#comment=27-804824 that
> was put by this #1310 <https://github.com/jenkinsci/jenkins/pull/1310>
> https://github.com/jenkinsci/jenkins/blob/master/.editorconfig#L16
> As i see this format sucks on options and my proposal to remove it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/1606#issuecomment-85392925>.
>

KostyaSha commented:
https://github.com/jenkinsci/jenkins/pull/1619
